cpp/ql/src/Likely Bugs/OO/UnsafeUseOfThis.ql

```diff
@@ -0,0 +1,288 @@
+/**
+ * @name Unsafe use of this in constructor
+ * @description A call to a pure virtual function using a this
+ *              pointer of an object that is under construction
+ *              may lead to undefined behavior.
+ * @kind path-problem
+ * @id cpp/unsafe-use-of-this
+ * @problem.severity error
+ * @precision high
+ * @tags correctness
+ *       language-features
+ *       security
+ */
+
+import semmle.code.cpp.ir.IR
+import cpp
+
+predicate irBbFunctionExit(IRBlock exit) {
+  exit.getLastInstruction() instanceof ExitFunctionInstruction
+}
+
+predicate irBbNodePred(IRBlock src, IRBlock pred) { src.getAPredecessor() = pred }
+
+predicate irBbIPostDominates(IRBlock postDominator, IRBlock node) =
+  idominance(irBbFunctionExit/1, irBbNodePred/2)(_, postDominator, node)
+
+predicate irBbStrictlyPostDominates(IRBlock postDominator, IRBlock node) {
+  irBbIPostDominates+(postDominator, node)
+}
+
+/**
+ * Holds if `postDominator` is a post-dominator of `node` in the control-flow graph. This
+ * is reflexive.
+ */
+predicate irBbPostDominates(IRBlock postDominator, IRBlock node) {
+  irBbStrictlyPostDominates(postDominator, node) or postDominator = node
+}
+
+bindingset[n, result]
+int unbind(int n) { result >= n and result <= n }
+
+/** Holds if `p` is the `n`'th parameter of function `f`. */
+predicate parameterOf(Parameter p, Function f, int n) { p.getFunction() = f and p.getIndex() = n }
```

Comment:
Is this the same as `f.getParameter(n) = p`?

Reply:
It should be, yes. Assuming the join-orderer is happy I like your version better!

Reply:
Done in e1ef41bd38af5924d0a4f3cab1267f75f77218e2.